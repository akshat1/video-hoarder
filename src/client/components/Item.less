@import (reference) "../style/vars.less";

.wrapper {
  .border;
  display: grid;
  grid-template-columns: auto;
  column-gap: @space-small;
  row-gap: @space-small;
  grid-template-areas:
    "title title"
    "statusIndicator url";
}

.label {
  .desktop-only;
  font-weight: bold;
  @media @desktop {
    &:after {
      content: ":"
    }
  }
}

.label, .value {
  @media @desktop {
    display: inline-block;
  }
}

.value {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  @media @desktop {
    overflow: unset;
    padding-left: @space-small;
  }
}

.title {
  grid-area: "title";
}

.url {
  grid-area: "url";
}
