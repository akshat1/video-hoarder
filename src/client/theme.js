import { createMuiTheme } from '@material-ui/core/styles';

const Theme = createMuiTheme({});

export { Theme };
