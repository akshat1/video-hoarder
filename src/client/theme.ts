import { createTheme } from "@material-ui/core/styles";
import _ from "lodash";

export const getTheme:any = _.memoize(darkMode =>
  createTheme({
    palette: {
      type: darkMode ? "dark" : "light",
    },
  }),
);
