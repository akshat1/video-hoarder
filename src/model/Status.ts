export enum Status {
  Failed = "failed",
  Paused = "paused",
  Pending = "pending",
  Running = "running",
  Succeeded = "succeeded",
}

export enum StatusFilterValue {
  All = "all",
  Failed = "failed",
  Paused = "paused",
  Pending = "pending",
  Running = "running",
  Succeeded = "succeeded",
}


export const hasStarted = (status: Status): boolean =>
  status === Status.Running || status === Status.Failed || status === Status.Succeeded || status === Status.Paused;

export const hasConcluded = (status: Status): boolean =>
  status === Status.Failed || status === Status.Succeeded;

export const isPending = (status: Status): boolean =>
  status === Status.Pending;
