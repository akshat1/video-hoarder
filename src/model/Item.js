/**
 * @module model/Item
 */

import { Status } from "../Status.js";
import md5 from "blueimp-md5";
import PropTypes from "prop-types";

/**
 * @typedef {Object} ItemMetadata
 */

/**
 * @typedef Item
 * @property {ItemMetadata} metadata
 * @property {Status} status
 * @property {string} addedBy - username of the user
 * @property {string} errorMessage
 * @property {string} id
 * @property {string} url
 * @property {TimeStamp} addedAt
 * @property {TimeStamp} updatedAt
 */

export const makeItem = (args) => {
  const {
    addedBy,
    url,
  } = args;

  const currentTime = new Date();
  const addedAt = currentTime.toISOString();

  return {
    addedAt,
    addedBy,
    id: md5(`${url}-${currentTime.getTime()}`),
    metadata: null,
    status: Status.Pending,
    updatedAt: addedAt,
    url,
  };
};

/**
 * For use in React prop-type declarations.
 * @const {Object}
 */
export const ItemShape = PropTypes.shape({
  addedAt: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  metadata: PropTypes.object,
  status: PropTypes.string.isRequired,
  thumbnail: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  updatedAt: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
});

/**
 * 
 * @param {Item} item 
 * @returns {string} - video title, or URL (as a fallback).
 */
export const getTitle = item => (item.metadata && item.metadata.title) || item.url;

/**
 * 
 * @param {Item} item 
 * @returns {string} - thumbnail URL or null.
 */
export const getThumbnail = item => (item.metadata && item.metadata.thumbnail) || null;

/**
 * 
 * @param {Item} item 
 * @returns {string} - video description.
 */
export const getDescription = item => (item.metadata && item.metadata.description);
