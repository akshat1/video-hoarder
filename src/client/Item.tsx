import { Job, JobStatus } from "../model/Job";
import { infoTable } from "./cssUtils";
import { Mutation } from "./gql";
import { JobProgress } from "./JobProgress";
import { Thumbnail } from "./Thumbnail";
import { useMutation } from "@apollo/client";
import { Button, Paper, Theme, Typography } from "@mui/material";
import { makeStyles } from "@mui/styles";
import _ from "lodash";
import React, { FunctionComponent } from "react";

const useStyle = makeStyles((theme: Theme) => ({
  root: {
    display: "grid",
    gridTemplateColumns: "1fr",
    gap: `${theme.spacing(1)} ${theme.spacing(1)}`,
    [theme.breakpoints.up("md")]: {
      gridTemplateColumns: "33% 1fr",
    },
  },
  thumbnail: {},
  thumbNailImage: {
    verticalAlign: "bottom",
  },
  notThumbnail: {
    padding: theme.spacing(1),
    display: "grid",
    gridTemplateRows: "min-content 1fr min-content",
    alignItems: "baseline",
  },
  title: {
    whiteSpace: "nowrap",
    overflow: "hidden",
    textOverflow: "ellipsis",
    width: "99.99%",
  },
  tableGrid: infoTable(theme),
  controls: {},
  progress: {
    marginTop: theme.spacing(2),
    marginBottom: theme.spacing(2),
  },
}));

interface Props {
  job: Job
}

export const Item:FunctionComponent<Props> = (props) => {
  const { job } = props;
  const classes = useStyle();

  const [
    doCancelJob, {
      error: cancellationError,
      loading: cancelling,
    },
  ] = useMutation(Mutation.CancelJob);
  const onCancel = () =>
    doCancelJob({
      variables: {
        jobId: job.id,
      },
    });

  const [
    doRemoveJob, {
      error: removalError,
      loading: removing,
    },
  ] = useMutation(Mutation.RemoveJob);
  const onRemove = () =>
    doRemoveJob({
      variables: {
        jobId: job.id,
      },
    });

  let updatedElem = null;
  if (job.updatedAt) {
    updatedElem = (
      <>
        <Typography>Last updated</Typography>
        <Typography>{new Date(job.updatedAt).toLocaleString(Intl.NumberFormat().resolvedOptions().locale)}</Typography>
      </>
    );
  }

  const opButtons = [];
  if (job.status === JobStatus.Pending || job.status === JobStatus.InProgress) {
    opButtons.push(
      <Button
        size="small"
        variant="outlined"
        onClick={onCancel}
        disabled={Boolean(cancelling || cancellationError)}
        key={`${job.id}--cancel`}
      >
        Cancel
      </Button>
    );
  }

  if (job.status === JobStatus.Canceled || job.status === JobStatus.Completed || job.status === JobStatus.Failed) {
    opButtons.push(<Button
      size="small"
      variant="outlined"
      onClick={onRemove}
      disabled={Boolean(removing || removalError)}
      key={`${job.id}--remove`}
    >
      Remove
    </Button>);
  }

  return (
    <Paper variant="outlined" className={classes.root}>
      <div className={classes.thumbnail}>
        <Thumbnail
          thumbnails={job.metadata.thumbnails}
          fallback={job.metadata.thumbnail}
          className={classes.thumbNailImage}
        />
      </div>
      <div className={classes.notThumbnail}>
        <Typography
          variant="h6"
          className={classes.title}
        >
          {job.metadata.title}
        </Typography>
        <div className={classes.tableGrid}>
          <Typography>Status</Typography>
          <Typography>{_.capitalize(job.status)}</Typography>
          <Typography>Download added</Typography>
          <Typography>{new Date(job.createdAt).toLocaleString(Intl.NumberFormat().resolvedOptions().locale)}</Typography>
          {updatedElem}
          <Typography>Format</Typography>
          <Typography>{_.get(job, "downloadOptions.formatSelector")}</Typography>
          <Typography>Rate limit</Typography>
          <Typography>{_.get(job, "downloadOptions.rateLimit")}</Typography>
          <Typography>Downloading to</Typography>
          <Typography>{_.get(job, "downloadOptions.downloadLocation")}</Typography>
        </div>
        <div className={classes.progress}>
          <JobProgress job={job} />
        </div>
        <div className={classes.controls}>
          {opButtons}          
        </div>
      </div>
    </Paper>
  );
};
