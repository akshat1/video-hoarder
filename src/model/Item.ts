import { Status } from "../model/Status";
import { Entity } from "./Entity";
import md5 from "blueimp-md5";

const System = "System";

export interface ItemMetadata {
  description: string,
  thumbnail: string,
  title: string,
}

export interface Item extends Entity {
  errorMessage?: string,
  metadata: ItemMetadata,
  status: Status,
  url: string,
}

export const makeItem = (args: { createdBy: string, url: string}): Item => {
  const {
    createdBy,
    url,
  } = args;

  const currentTime = new Date();
  const createdAt = currentTime.toISOString();
  return {
    createdAt,
    createdBy,
    id: md5(`${url}-${currentTime.getTime()}`),
    metadata: null,
    status: Status.Pending,
    updatedAt: createdAt,
    updatedBy: createdBy,
    url,
  };
};

// Accessors
export const getTitle = (item: Item): string => item?.metadata?.title || item.url;

export const getThumbnail = (item: Item): string => item?.metadata?.thumbnail || null;

export const getDescription = (item: Item): string => item?.metadata?.description || "";

// Mutators

/**
 * Sets the status to Failed.
 */
export const markItemFailed = (args: { errorMessage: string, item: Item }): Item => {
  const { errorMessage, item } = args;
  return {
    ...item,
    errorMessage,
    status: Status.Failed,
    updatedAt: new Date().toISOString(),
    updatedBy: System,
  };
};

/**
 * Sets the status to Canceled (a particular failure mode with an associated user).
 */
export const markItemCanceled = (args: { item: Item, updatedBy: string }): Item => {
  const { item, updatedBy } = args;
  return {
      ...item,
    errorMessage: "Canceled",
    status: Status.Failed,
    updatedAt: new Date().toISOString(),
    updatedBy,
  }
};

/**
 * Sets the status to Succesful.
 * @returns a new Item object with the updated status.
 */
export const markItemSuccessful = (item:Item): Item => ({
  ...item,
  status: Status.Succeeded,
  updatedAt: new Date().toISOString(),
  updatedBy: System,
});

/**
 * Sets the status to InProgress.
 * @returns a new Item object with the updated status.
 */
export const markItemInProgress = (item: Item): Item => ({
  ...item,
  status: Status.Running,
  updatedAt: new Date().toISOString(),
  updatedBy: System,
});

/**
 * Sets Metadata.
 * @returns a new Item object with the updated metadata.
 */
export const setMetadata = (args: { item, metadata }): Item => {
  const { item, metadata } = args;
  return {
    ...item,
    updatedAt: new Date().toISOString(),
    metadata,
  };
};
