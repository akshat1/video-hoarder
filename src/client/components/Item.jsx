/**
 * Renders a single download-task, and displays metadata and status of the same.
 * Also provides controls to abort the download.
 */
import { getDescription, getThumbnail, getTitle, ItemShape } from "../../model/Item.js";
import CancelButton from "./CancelButton.jsx";
import ItemMeta from "./ItemMeta.jsx";
import { Button, Collapse, Grid, Link, Typography, useMediaQuery, } from "@material-ui/core";
import ExpandMoreIcon from "@material-ui/icons/ExpandMore";
import { makeStyles, useTheme, } from "@material-ui/styles";
import ANSIToHTML from "ansi-to-html";
import classnames from "classnames";
import React from "react";

const convert = new ANSIToHTML({
  newline: true,
});

const useStyle = makeStyles(theme => {
  console.log(theme);
  return {
    root: {
      padding: theme.spacing(1),
    },

    thumbnail: {
      width: "100%",
    },

    expandIcon: {
      transform: "rotate(0deg)",
      transition: theme.transitions.create("transform", {
        duration: theme.transitions.duration.shortest,
      }),
    },

    expandOpen: {
      transform: "rotate(180deg)",
    },

    title: {
      fontSize: "1.25em",
      [theme.breakpoints.up("md")]: {
        fontSize: "2em"
      }
    },
    url: {},
    cancelButton: {
      marginTop: theme.spacing(2),
      [theme.breakpoints.down("xs")]: {
        float: "right"
      }
    }
  };
});

const Item = (props) => {
  const classes = useStyle();
  const { item } = props;
  const theme = useTheme();
  const isMobile = useMediaQuery(theme.breakpoints.down("xs"));
  const [expanded, setExpanded] = React.useState(false);

  const handleExpandClick = () => {
    setExpanded(!expanded);
  };

  const { url } = item;
  const title = getTitle(item);
  const thumbnail = getThumbnail(item);
  const mediaTitle = `${title} thumbnail`;
  const expandIcon = <ExpandMoreIcon
    className={classnames(classes.expandIcon, {
      [classes.expandOpen]: expanded,
    })}
  />;

  return (
    <Grid container className={classes.root} spacing={1}>
      <Grid item xs={4} md={2}>
        <img src={thumbnail} title={mediaTitle} className={classes.thumbnail} />
      </Grid>
      <Grid item xs={8} md={10}>
        <Typography className={classes.title} variant="h2">{title}</Typography>
        <If condition={!isMobile}>
          <ItemMeta item={item} />
          <Typography className={classes.url}>
            <Link href={url}>{url}</Link>
          </Typography>
          <CancelButton item={item} className={classes.cancelButton} />
        </If>
      </Grid>
      <If condition={isMobile}>
        <Grid item xs={12}>
          <ItemMeta item={item} />
        </Grid>
        <Grid item xs={12}>
          <CancelButton item={item} className={classes.cancelButton} />
        </Grid>
      </If>
      <Grid item xs={12}>
        <Button
          className={classes.expand}
          onClick={handleExpandClick}
          aria-expanded={expanded}
          aria-label="show more"
          endIcon={expandIcon}
        >
          <Typography className={classes.descriptionLabel} component="span">Description</Typography>
        </Button>
        <Collapse in={expanded} unmountOnExit>
          <If condition={isMobile}>
            <Typography className={classes.url}>
              <Link href={url}>{url}</Link>
            </Typography>
          </If>
          <Typography className={classes.description} dangerouslySetInnerHTML={{ __html: convert.toHtml(getDescription(item)) }} />
        </Collapse>
      </Grid>
    </Grid>
  );
};

Item.propTypes = {
  item: ItemShape,
};

export default Item;
