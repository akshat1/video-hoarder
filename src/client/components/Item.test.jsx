import React from 'react';
import { shallow } from 'enzyme';
import { ThemeProvider } from '@material-ui/styles';
import { getTheme } from '../theme';
import Item from './Item.jsx';
import { Status } from '../../Status';
import { getFakeItem } from '../fixtures/item';

describe('components/Item', () => {
  Object.values(Status).forEach((status) =>
    test(`Item matches snapshot for ${status}`, () => {
      const instance = <ThemeProvider theme={getTheme()}><Item item={getFakeItem(status)} />)</ThemeProvider>
      expect(shallow(instance)).toMatchSnapshot();
    })
  );
});
