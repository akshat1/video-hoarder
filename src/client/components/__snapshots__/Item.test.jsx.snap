// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Item Item matches snapshot for failed 1`] = `
<div
  className="wrapper failed"
>
  <div
    className="title"
  >
    A collectible video
  </div>
  <div
    className="url"
  >
    <div
      className="value"
      title="https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux"
    >
      <a
        href="https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux
      </a>
    </div>
  </div>
  <div
    className="status"
  >
    <ItemStatus
      status="failed"
    />
  </div>
  <div
    className="meta"
  >
    <ItemMeta
      item={
        Object {
          "addedAt": "2020-03-29T14:30:45.000Z",
          "id": "3idh834dh348dhhfufh48",
          "status": "failed",
          "title": "A collectible video",
          "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
          "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
        }
      }
    />
  </div>
</div>
`;

exports[`components/Item Item matches snapshot for pending 1`] = `
<div
  className="wrapper pending"
>
  <div
    className="title"
  >
    A collectible video
  </div>
  <div
    className="url"
  >
    <div
      className="value"
      title="https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux"
    >
      <a
        href="https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux
      </a>
    </div>
  </div>
  <div
    className="status"
  >
    <ItemStatus
      status="pending"
    />
  </div>
  <div
    className="meta"
  >
    <ItemMeta
      item={
        Object {
          "addedAt": "2020-03-29T14:30:45.000Z",
          "id": "3idh834dh348dhhfufh48",
          "status": "pending",
          "title": "A collectible video",
          "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
          "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
        }
      }
    />
  </div>
</div>
`;

exports[`components/Item Item matches snapshot for running 1`] = `
<div
  className="wrapper running"
>
  <div
    className="title"
  >
    A collectible video
  </div>
  <div
    className="url"
  >
    <div
      className="value"
      title="https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux"
    >
      <a
        href="https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux
      </a>
    </div>
  </div>
  <div
    className="status"
  >
    <ItemStatus
      status="running"
    />
  </div>
  <div
    className="meta"
  >
    <ItemMeta
      item={
        Object {
          "addedAt": "2020-03-29T14:30:45.000Z",
          "id": "3idh834dh348dhhfufh48",
          "status": "running",
          "title": "A collectible video",
          "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
          "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
        }
      }
    />
  </div>
</div>
`;

exports[`components/Item Item matches snapshot for succeeded 1`] = `
<div
  className="wrapper succeeded"
>
  <div
    className="title"
  >
    A collectible video
  </div>
  <div
    className="url"
  >
    <div
      className="value"
      title="https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux"
    >
      <a
        href="https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux
      </a>
    </div>
  </div>
  <div
    className="status"
  >
    <ItemStatus
      status="succeeded"
    />
  </div>
  <div
    className="meta"
  >
    <ItemMeta
      item={
        Object {
          "addedAt": "2020-03-29T14:30:45.000Z",
          "id": "3idh834dh348dhhfufh48",
          "status": "succeeded",
          "title": "A collectible video",
          "updatedAt": "Sun Mar 29 2020 14:32:30 GMT+0000 (Coordinated Universal Time)",
          "url": "https://www.youtube.com/watch?v=eQ5Ru7Zu_1I?foo=bar&baz=qux",
        }
      }
    />
  </div>
</div>
`;
